rcl/test/rcl/test_get_actual_qos.cpp

```diff
@@ -0,0 +1,181 @@
+    ret = rcl_init(0, nullptr, &init_options, this->context_ptr);
+    ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+    this->node_ptr = new rcl_node_t;
+    *this->node_ptr = rcl_get_zero_initialized_node();
+    const char * name = "test_get_actual_qos_node";
+    ret = rcl_node_init(this->node_ptr, name, "", this->context_ptr, &node_options);
+    ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  }
+
+  void TearDown()
+  {
+    rcl_ret_t ret;
+
+    ret = rcl_node_fini(this->node_ptr);
+    delete this->node_ptr;
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+
+    ret = rcl_shutdown(this->context_ptr);
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+
+    ret = rcl_context_fini(this->context_ptr);
+    delete this->context_ptr;
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  }
+};
+
+TEST_P(CLASSNAME(TestGetActualQoS, RMW_IMPLEMENTATION), test_publisher_get_qos_settings) {
+  CLASSNAME(TestParameters, RMW_IMPLEMENTATION) parameters =
+    GetParam();
+  std::string topic_name("/test_publisher_get_actual_qos__");
+  rcl_ret_t ret;
+
+  rcl_publisher_t pub = rcl_get_zero_initialized_publisher();
+  rcl_publisher_options_t pub_ops = rcl_publisher_get_default_options();
+  pub_ops.qos = parameters.qos_to_set;
+  auto ts = ROSIDL_GET_MSG_TYPE_SUPPORT(test_msgs, msg, Primitives);
+  ret = rcl_publisher_init(&pub, this->node_ptr, ts, topic_name.c_str(), &pub_ops);
+  EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  rcl_reset_error();
+
+  rmw_qos_profile_t qos;
+  ret = rcl_publisher_get_actual_qos(&pub, &qos);
+  EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  EXPECT_EQ(
+    parameters.qos_to_set.history,
+    parameters.qos_expected.history);
+  EXPECT_EQ(
+    parameters.qos_to_set.depth,
+    parameters.qos_expected.depth);
+  EXPECT_EQ(
+    parameters.qos_to_set.reliability,
+    parameters.qos_expected.reliability);
+  EXPECT_EQ(
+    parameters.qos_to_set.durability,
+    parameters.qos_expected.durability);
+  EXPECT_EQ(
+    parameters.qos_to_set.avoid_ros_namespace_conventions,
+    parameters.qos_expected.avoid_ros_namespace_conventions);
+
+  ret = rcl_publisher_fini(&pub, this->node_ptr);
+  EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  rcl_reset_error();
+}
+
+CLASSNAME(TestParameters, RMW_IMPLEMENTATION) parameters[] = {
+  /*
+    Testing with default qos settings.
+   */
+  {
+    rmw_qos_profile_default,
+    rmw_qos_profile_default,
+    "publisher_default_qos"
+  },
+  /*
+    Test with non-default settings.
+   */
+  {
+    {
+      RMW_QOS_POLICY_HISTORY_KEEP_ALL,
+      1000,
+      RMW_QOS_POLICY_RELIABILITY_BEST_EFFORT,
+      RMW_QOS_POLICY_DURABILITY_TRANSIENT_LOCAL,
+      true
+    },
+    {
+      RMW_QOS_POLICY_HISTORY_KEEP_ALL,
+      1000,
+      RMW_QOS_POLICY_RELIABILITY_BEST_EFFORT,
+      RMW_QOS_POLICY_DURABILITY_TRANSIENT_LOCAL,
+      true
+    },
+    "publisher_non_default_qos"
+  }
```

Comment:
Is there no way to test `rmw_qos_profile_system_default`? I guess not without a custom rmw implementation to test it.

Also not covered is the case where the underlying QoS setting cannot be represented by ROS QoS.

Reply:
I can do a test with "qos_to_set=rmw_qos_profile_default" and a custom qos_expected for each of the rmw implementations (using the preprocessor).
For getting an "unknown" response, I would need to access the DDS api directly. I'm ok with not covering it. If we want to cover it, I could work in a custom test for each of the vendors (or at least in one of them).

Reply:
Sounds good.